File: prop-core/src/main/java/io/pleo/prop/core/internal/LocalProp.java
package io.pleo.prop.core.internal;

import java.time.Instant;
import java.util.function.Supplier;

import io.pleo.prop.core.Prop;

public class LocalProp<T> implements Prop<T> {

  private final Supplier<T> supplier;
  private final Instant created;

  public LocalProp(T value) {
    this(() -> value);
  }

  public LocalProp(Supplier<T> supplier) {
    this.supplier = supplier;
    this.created = Instant.now();
  }

  @Override
  public T get() {
    return supplier.get();
  }

  @Override
  public String getName() {
    return toString();
  }

  @Override
  public Instant getChangedTimestamp() {
    return created;
  }

  @Override
  public void addCallback(Runnable callback) {
    // Static properties can't change. No callbacks.
  }

  @Override
  public void removeAllCallbacks() {
    // Static properties can't change. No callbacks.
  }

  @Override
  public String toString() {
    return "LocalProp=" + get();
  }
}

File: prop-core/src/main/kotlin/io/pleo/prop/core/internal/LocalProp.kt
package io.pleo.prop.core.internal

import io.pleo.prop.core.Callback
import io.pleo.prop.core.Prop
import io.pleo.prop.core.Supplier
import java.time.Instant

class LocalProp<T>(
    private val supplier: Supplier<T>,
) : Prop<T> {
    private val created: Instant = Instant.now()

    constructor(value: T) : this({ value })

    override fun invoke() = supplier()

    override fun get() = invoke()

    override val name get() = toString()
    override val changedTimestamp get() = created

    /** Static properties can't change. No callbacks.*/
    override fun addCallback(callback: Callback<T>) = Unit

    override fun removeAllCallbacks() = Unit

    override fun toString() = "LocalProp=${get()}"
}
